File: nvpl_blas/f77/ztrmv.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     ZTRMV 
!
!******************************************************************************
      program   ZTRMV_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

*       Intrinsic Functions
      intrinsic  abs

      character*1     uplo, transa, diag
      parameter       (uplo='U', transa='N', diag='N')
      integer         n, lda
      integer         incx, i
      parameter       (n=2, lda=2, incx=1)
      double complex  a(lda,n)
      integer         xsize
      parameter       (xsize=1+(n-1)*abs(incx))
      double complex  x(xsize)
      data            (x(i), i=1,xsize)/1.0,-3.7/

*       External Subroutines
      external        print_zmatrix, fill_zmatrix, print_zvector
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, n, lda, incx, uplo, transa, diag
      call fill_zmatrix(a,n,n,lda,uplo,diag)
      call print_zmatrix(a,n,n,lda,'A')
      call print_zvector(x,n,incx,'X')
*
*      Call ZTRMV subroutine
      call ZTRMV(uplo,transa,diag,n,a,lda,x,incx)
*
      print*
      print 102, 'After ztrmv operation:'
      call print_zvector(x,n,incx,'X')

      stop
 99   format('Example: ZTRMV for the triangular matrix-vector '
     &       'multiplication')
 100  format('#### args: n=',i1,', lda=',i1,', incx=',i1,', uplo=',a1,
     &       ', transa=',a1,', diag=',a1)
 102  format(A)
      end

File: nvpl_blas/c/ztrmv.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_ztrmv
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t N = 2;
    nvpl_int_t lda = 2;
    enum CBLAS_ORDER order = CblasRowMajor;
    enum CBLAS_TRANSPOSE transA = CblasNoTrans;
    enum CBLAS_UPLO uplo = CblasUpper;
    enum CBLAS_DIAG diag = CblasNonUnit;
    nvpl_dcomplex_t * A = NULL;
    nvpl_dcomplex_t * X;
    nvpl_int_t incX = 1;

    printf("\nExample: cblas_ztrmv for the triangular matrix-vector multiplication\n\n");
    printf("#### args: n=%" PRId64 ", lda=%" PRId64 ", incx=%" PRId64 ", transA=%c, order=%c, uplo=%c"
            ", diag=%c\n", (int64_t)N, (int64_t)lda, (int64_t)incX, transpose_to_char(transA), 
            order_to_char(order), uplo_to_char(uplo), diag_to_char(diag));

    nvpl_int_t len_x = 1 + (N - 1) * labs(incX);

    // allocate memory
    A = (nvpl_dcomplex_t *)malloc(lda * N * sizeof(nvpl_dcomplex_t));
    X = (nvpl_dcomplex_t *)malloc(len_x * sizeof(nvpl_dcomplex_t));

    // fill data
    fill_zmatrix(A, N, N, lda, order, UPLO_TO_FILL_MODE(uplo), diag);
    fill_zvector(X, N, incX);

    // print input data
    print_zmatrix(A, N, N, lda, "A", order);
    print_zvector(X, N, incX, "X");

    // call cblas_ztrmv
    cblas_ztrmv(order, uplo, transA, diag, N, A, lda, X, incX);

    // print result
    printf("\nAfter ztrmv operation:\n");
    print_zvector(X, N, incX, "X");

    // release memory
    free(A);
    free(X);
    return 0;
}
